app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.kt

```diff
@@ -0,0 +1,746 @@
+import org.schabi.newpipe.util.NavigationHelper
+import org.schabi.newpipe.util.ServiceHelper
+import java.util.stream.Collectors
+
+class MediaBrowserConnector(
+    private val playerService: PlayerService,
+) : PlaybackPreparer {
+    private val mediaSession = MediaSessionCompat(playerService, TAG)
+    val sessionConnector = MediaSessionConnector(mediaSession).apply {
+        setMetadataDeduplicationEnabled(true)
+        setPlaybackPreparer(this@MediaBrowserConnector)
+    }
+
+    private val database: AppDatabase by lazy { NewPipeDatabase.getInstance(playerService) }
+    private val localPlaylistManager: LocalPlaylistManager by lazy { LocalPlaylistManager(database) }
+    private val remotePlaylistManager: RemotePlaylistManager by lazy {
+        RemotePlaylistManager(database)
+    }
+    private val playlists: Flowable<List<PlaylistLocalItem?>>
+        get() {
+            return MergedPlaylistManager.getMergedOrderedPlaylists(localPlaylistManager, remotePlaylistManager)
+        }
+
+    private var prepareOrPlayDisposable: Disposable? = null
+    private var searchDisposable: Disposable? = null
+    private var bookmarksNotificationsDisposable: Disposable? = null
+
+    init {
+        playerService.sessionToken = mediaSession.sessionToken
+        setupBookmarksNotifications()
+    }
+
+    fun release() {
+        disposePrepareOrPlayCommands()
+        disposeBookmarksNotifications()
+        mediaSession.release()
+    }
+
+    private fun createRootMediaItem(
+        mediaId: String?,
+        folderName: String,
+        @DrawableRes iconResId: Int
+    ): MediaBrowserCompat.MediaItem {
+        val builder = MediaDescriptionCompat.Builder()
+        builder.setMediaId(mediaId)
+        builder.setTitle(folderName)
+        val resources = playerService.resources
+        builder.setIconUri(
+            Uri.Builder()
+                .scheme(ContentResolver.SCHEME_ANDROID_RESOURCE)
+                .authority(resources.getResourcePackageName(iconResId))
+                .appendPath(resources.getResourceTypeName(iconResId))
+                .appendPath(resources.getResourceEntryName(iconResId))
+                .build()
+        )
+
+        val extras = Bundle()
+        extras.putString(
+            MediaConstants.DESCRIPTION_EXTRAS_KEY_CONTENT_STYLE_GROUP_TITLE,
+            playerService.getString(R.string.app_name)
+        )
+        builder.setExtras(extras)
+        return MediaBrowserCompat.MediaItem(
+            builder.build(),
+            MediaBrowserCompat.MediaItem.FLAG_BROWSABLE
+        )
+    }
+
+    private fun createPlaylistMediaItem(playlist: PlaylistLocalItem): MediaBrowserCompat.MediaItem {
+        val builder = MediaDescriptionCompat.Builder()
+        val isRemote = playlist is PlaylistRemoteEntity
+        builder.setMediaId(createMediaIdForInfoItem(isRemote, playlist.uid))
+            .setTitle(playlist.orderingName)
+            .setIconUri(Uri.parse(playlist.thumbnailUrl))
```

Comment:
Does this work if `playlist.thumbnailUrl == null`?

Reply:
You are right. Added null check here.